[llvm][transforms] Add a new algorithm to SplitModule

sergachev opened the pull request:
The new round-robin algorithm overrides the hash-based distribution of functions to modules. It achieves a more even number of functions per module when the number of functions is close to the number of requested modules. It's not in use by default and is available under a new flag.

sergachev committed:
[Transforms] Add a new algorithm to SplitModule

The new algorithm augments the hash-based distribution of functions to
modules by pairing unmapped functions to empty modules before using the
hash-based distribution. It allows getting less empty modules when the
number of functions is close to the number of requested modules.
It's not in use by default and is available under a new flag.

grypp commented on llvm/tools/llvm-split/llvm-split.cpp:
```diff
@@ -53,6 +53,12 @@ static cl::opt<bool>
                    cl::cat(SplitCategory));
 
+static cl::opt<bool>
+    TryToAvoidEmptyModules("avoid-empty-modules", cl::Prefix, cl::init(false),
```
I am wondering can we make this option default? We could still introduce a flag that opt-outs?

sergachev replied:
Can be done but I'd prefer changing the default behavior to be a separate change so that the current change would not need to be reverted if someone relies on the current behavior.

grypp reviewed (commented):
I am not expert here, but it looks good to me.
Nit: title should be **[llvm][transforms]**

fhahn commented:
Could you share some statistics on how this performs on larger inputs? (e.g. llvm-test-suite)

sergachev commented:
> Could you share some statistics on how this performs on larger inputs? (e.g. llvm-test-suite)

Sure! https://gist.github.com/sergachev/f68e5f6dad591d1efbcfa3fcd23b140c#file-results-txt-L4-L41

sergachev commented:
ping

joker-eph commented on llvm/lib/Transforms/Utils/SplitModule.cpp:
```diff
@@ -268,6 +268,38 @@ void llvm::SplitModule(
+        break;
+      ClusterIDMap.insert({F, *NextEmptyModuleIt});
+      ++NextEmptyModuleIt;
+    }
```
I don't quite get this loop, seems like you won't distribute all the unmapped functions, but only at most a number equal to the empty modules? All the others will be distributed by isInPartition() if I read this loop correctly.

Why don't you just circle back to `EmptyModules.begin()` instead of breaking out of the loop?

Also should we skip the loop over UnmappedFunctions if `EmptyModules.empty()` ?

sergachev replied:
> I don't quite get this loop, seems like you won't distribute all the unmapped functions, but only at most a number equal to the empty modules? All the others will be distributed by isInPartition() if I read this loop correctly.

Correct. Our main use case for this is to request the number of modules equal to the number of functions - then (provided that the functions are independent) the algorithm generates exactly 1 function per module. The behavior corresponds to the name of the flag - TryToAvoidEmptyModules.

> Why don't you just circle back to EmptyModules.begin() instead of breaking out of the loop?

Can do that too - this would become a different algorithm, completely overriding the hash-based distribution. For our goal (1 function per module) this would work equally. Do you think we should change it this way?

> Also should we skip the loop over UnmappedFunctions if EmptyModules.empty() ?

Indeed. Fixed.

joker-eph replied:
> Our main use case for this is to request the number of modules equal to the number of functions

I see: that is quite specific, it also makes the algorithm (and flags) only really suited for this case it seems. The name  "TryToAvoidEmptyModules" is explicit, but one may wonder "why would anyone use this algorithm as is?" without knowing the implicit assumption about "the number of modules equal to the number of functions".

sergachev replied:
>"why would anyone use this algorithm as is?"

To get less useless empty modules.

> only really suited for this case it seems

It's best suited for this use case but works fine in the other cases too and does the same - reduces the number of empty modules as compared to what hash-based distribution does.

joker-eph replied:
> To get less useless empty modules.
>  reduces the number of empty modules as compared to what hash-based distribution does.

Right, I know it'll "work", but "to get less useless empty modules." isn't an obvious optimization metrics to me! It just seems like an odd incomplete solution to a load-balancing problem.

sergachev replied:
As suggested I extended the new algorithm to do round-robin of all unmapped functions completely overriding the name-hash-based distribution. This works the same way for our main use case (number of modules equal to the number of functions) and may be more useful in other use cases. In the updated test this produces 2+2 function modules instead of the default 1+3.

sergachev committed:
Skip redundant work if EmptyModules is empty

address review comment

sergachev committed:
Document flags of SplitModule()

address review comment

sergachev commented on llvm/include/llvm/Transforms/Utils/SplitModule.h:
```diff
@@ -35,7 +35,7 @@ class Module;
     Module &M, unsigned N,
     function_ref<void(std::unique_ptr<Module> MPart)> ModuleCallback,
-    bool PreserveLocals = false);
+    bool PreserveLocals = false, bool TryToAvoidEmptyModules = false);
```
done

sergachev committed:
Extend the algorithm to do round-robin distribution

completely override the hash-based distribution by mapping all unmapped
functions to modules using round-robin

joker-eph reviewed (approved):
LG, thanks!
